engine/src/flutter/display_list/dl_builder.cc

```diff
@@ -369,17 +370,17 @@ void DisplayListBuilder::SetAttributesFromPaint(
     setStrokeJoin(paint.getStrokeJoin());
   }
   if (flags.applies_shader()) {
-    setColorSource(paint.getColorSource().get());
+    setColorSource(paint.getColorSourcePtr());
```

Comment:
Use the ptr APIs directly so we don't increment + decrement the shared_ptr

Reply:
Doesn't it have a shared_ptr interface to avoid the copy? Wouldn't that be better?

Reply:
I don't think it matters since we're not storing anything.

Reply:
The issue here is that setColorSource can call onSetColorSource if the objects are different which needs a shared object. This should have some more refactoring otherwise we are recreating the object inside onSetColorSource.

We need setColorSource(ptr) because builder implements DlOpReceiver for copying of data. I might be able to get rid of that eventually, but until then perhaps the SetAttriutesFrom method should do its own comparisons and onSetColorSource takes a shared_ptr (ref). Thus, the DlOpReceiver version might need to call shared(), but not the one that syncs with a paint. Another issue is that if it is null then you can't call shared(). :(

Reply:
hmm, yeah we want to avoid calling shared_from_this if possible since that does a weak ptr lock. Instead I'll change the setCall to take a refer and then we can avoid the shared call when non-null.

Reply:
There is no shared_from_this currently. We have shared() which instantiates a whole new copy inside a shared_ptr. It's even worse than a weak ptr lock...?

Reply:
eh, then it will be fine. I could keep refactoring this for a while but since I'm about to go OOO maybe call this good enough?

Reply:
I have no objections and @gaaclarke already approved it. I think these changes move us forward.

Reply:
ty!